### PR TITLE
feat: enable generic claims for RLS policies

### PR DESCRIPTION
**Pull request by w3b6x9**
## What kind of change does this PR introduce?

feature

## What is the new behavior?

Subscribers can now pass the entire JWT claims to Realtime server, enabling developers to create different database roles and create RLS policies that include those roles and/or anything else inside the JWT claims.

## Additional context

Add any other context or screenshots.

- https://github.com/supabase/walrus/pull/30
- https://github.com/supabase/gotrue/pull/289 and https://github.com/supabase/gotrue/pull/295

### CODE REVIEW

File: server/priv/repo/migrations/20211210212804_enable_generic_subscription_claims.exs [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+    execute "alter table realtime.subscription rename user_id to subscription_id"
+
+    execute "create function realtime.to_regrole(role_name text)
+      returns regrole
+      immutable
+      language sql
+      -- required to allow use in generated clause
+      as $$ select role_name::regrole $$;"
+
+    execute "alter table realtime.subscription
```
Comment (olirice) [inline server/priv/repo/migrations/20211210212804_enable_generic_subscription_claims.exs:18]: the migration fail on this line with 

```
ERROR:  column "claims" of relation "subscription" contains null values
SQL state: 23502
```

if there are rows in `realtime.subscription`

Should there be a `truncate table realtime.subscription` at the top of the migration?

**Review by abc3 — APPROVED** [review]
Looks good, only check rows ordering in ReplicationPoller

File: server/priv/repo/migrations/20211210212804_enable_generic_subscription_claims.exs [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+      add column claims jsonb not null,
+      add column claims_role regrole not null generated always as (realtime.to_regrole(claims ->> 'role')) stored"
```
Comment (olirice) [inline server/priv/repo/migrations/20211210212804_enable_generic_subscription_claims.exs:20]: would be good to drop the `created_at` column and recreate it here to keep the ordering consistent

File: server/priv/repo/migrations/20211210212804_enable_generic_subscription_claims.exs (hunk 1 above)
Comment (w3b6x9) [reply server/priv/repo/migrations/20211210212804_enable_generic_subscription_claims.exs:18]: @olirice good call! will add it in

File: server/priv/repo/migrations/20211210212804_enable_generic_subscription_claims.exs (hunk 2 above)
Comment (w3b6x9) [reply server/priv/repo/migrations/20211210212804_enable_generic_subscription_claims.exs:20]: @olirice sounds good, will do!

File: server/lib/realtime/rls/replication_poller.ex [hunk 3]
```diff
@@ -86,19 +86,24 @@ defmodule Realtime.RLS.ReplicationPoller do
...
+            |> generate_record()
+            |> case do
+              nil ->
+                acc
+
+              record_struct ->
+                [record_struct | acc]
+            end
+          end)
+          |> SubscribersNotification.notify()
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply server/lib/realtime/rls/replication_poller.ex:106]: @abc yes, good call!

**Pull request merged**